### PR TITLE
solver: swap catas with recursion, fuse travs

### PR DESCRIPTION
**Pull request by gbaz**
This is the last refactor I had in mind to address the general performance issues discussed in https://github.com/haskell/cabal/issues/7466

It attains some modest but real improvement (mainly in allocations) from eliminating the intermediate `cata` calls in favor of explicit recursion in the solver tree.

For `trav` calls, rather than eliminate them entirely, it lifts out their intermediate functions to the top level and explicitly fuses them in the solver pipeline, which yields somewhat more significant improvements.

Overall not the huge win I hoped, but nonetheless given the straightforward nature of the changes to the code, still worth it I think.

### CODE REVIEW

File: cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs [hunk 1]
```diff
@@ -164,12 +164,12 @@ type VROrigin = (VR, ExposedComponent, DependencyReason QPN)
...
 validate :: Tree d c -> Validate (Tree d c)
-validate = cata go
+validate = go
   where
-    go :: TreeF d c (Validate (Tree d c)) -> Validate (Tree d c)
+    go :: Tree d c -> Validate (Tree d c)
 
-    go (PChoiceF qpn rdm gr       ts) = PChoice qpn rdm gr <$> sequence (W.mapWithKey (goP qpn) ts)
-    go (FChoiceF qfn rdm gr b m d ts) =
+    go (PChoice qpn rdm gr       ts) = PChoice qpn rdm gr <$> sequence (W.mapWithKey (goP qpn) (fmap go ts))
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs:171]: sequence . map seems bad idea. Should we add `traverseWithKey` to `WeightedPSQ` (separately)? Maybe it fuses, but  I'm somewhat sceptical about that.

Comment (gbaz) [reply cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs:171]: Sure. I take it you don't consider that a blocker for this PR, as its about changing a different thing than this PR currently does. Are you ok with approving this pr as is?

**Review by emilypi — APPROVED** [review]

File: cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs (hunk 1 above)
Comment (emilypi) [reply cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs:171]: I can fix it up with `traverseWithKey` in the follow up #7533

Comment (emilypi) [reply cabal-install-solver/src/Distribution/Solver/Modular/Validate.hs]: Added it first, then pushed to this branch.

**Comment by emilypi** [comment]
https://github.com/haskell/cabal/pull/7519

**Comment by emilypi** [comment]
@mergifyio backport 3.6

**Pull request merged**